File: cucumber-kotlin/src/test/kotlin/api/SUT.kt
package api

import io.restassured.RestAssured
import io.restassured.RestAssured.given
import io.restassured.RestAssured.`when`
import io.restassured.http.ContentType
import io.restassured.response.Response

data class Coordinates(val lat: Int, val lon: Int)

class SUT {
    companion object {
        init {
            RestAssured.baseURI = "http://localhost:3000"
        }

        fun addLocation(name: String, coordinates: Coordinates): Response {
            val jsonAsMap = mapOf(
                "name" to name,
                "coordinates" to mapOf(
                    "lat" to coordinates.lat,
                    "lon" to coordinates.lon
                )
            )
            return given()
                    .contentType(ContentType.JSON)
                    .body(jsonAsMap)
                .`when`()
                    .post("/locations")
        }

        fun deleteLocation(locationName: String): Response {
            return given()
                    .pathParam("locationName", locationName)
                .`when`()
                    .delete("/locations/{locationName}")
        }

        fun getLocations(): Response {
            return `when`().get("/locations")
        }

        fun getLocation(locationName: String): Response {
            return given()
                    .pathParam("locationName", locationName)
                .`when`()
                    .get("/locations/{locationName}")
        }

        fun getHealth(): Response {
            return `when`().get("/health")
        }

        fun getYrHealth(): Response {
            return `when`().get("/health/yr")
        }

        fun getWeather(locationName: String): Response {
            return given()
                    .pathParam("locationName", locationName)
                .`when`()
                    .get("/weather/{locationName}")

        }
    }
}

File: cucumber/step-definitions/api/sut.ts
const SUT_BASE_URL = 'http://localhost:3000'

export interface Response {
    status: number,
    statusText: string,
    body: any
}

export interface Coordinates {
    lat: number,
    lon: number
}

export let response: Response | undefined

async function doRequest(url: string, reqParams: RequestInit,
    urlParams: URLSearchParams | undefined = undefined): Promise<Response> {

    let reqUrl = url
    if (urlParams !== undefined) {
        reqUrl += '?'+ urlParams
    }
    console.log('Request to SUT --->')
    console.log('Request URL:', reqUrl)
    console.log('Request parameters:', reqParams)
    const response = await fetch(reqUrl, reqParams);
    const responseStatus = response.status + ' ' + response.statusText

    console.log('Response status:', responseStatus)
    let responseBody = undefined
    if (reqParams.method !== 'DELETE') {
        try {
            responseBody = await response.json()
            // console.log('Response body:', JSON.stringify(responseBody, null, 4));
        }
        catch(e) {
            console.log('Response body:', e)
        }
    }
    console.log('<---')

    return { 
        status: response.status, 
        statusText: response.statusText,
        body: responseBody
    }
}

export async function addLocation(name: string, coordinates: Coordinates) {
    return await doRequest(SUT_BASE_URL + '/locations', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        },
        body: JSON.stringify({ name: name, coordinates: coordinates })
    })
}

export async function deleteLocation(locationName: string) {
    return await doRequest(SUT_BASE_URL + '/locations/' + locationName, {
        method: 'DELETE',
        headers: {
            'Accept': 'application/json'
        }
    })
}

export async function getLocations() {
    return await doRequest(SUT_BASE_URL + '/locations', {
        method: 'GET',
        headers: {
            'Accept': 'application/json'
        }
    })
}

export async function getLocation(locationName: string) {
    return await doRequest(SUT_BASE_URL + '/locations/' + locationName, {
        method: 'GET',
        headers: {
            'Accept': 'application/json'
        }
    })
}

export async function getHealth() {
    return await doRequest(SUT_BASE_URL + '/health', {
            method: 'GET',
            headers: {
                'Accept': 'application/json'
            }
        }
    )
}

export async function getYrHealth() {
    return await doRequest(SUT_BASE_URL + '/health/yr', {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
            }
        }
    )
}

export async function getWeather(locationName: string) {
    return await doRequest(SUT_BASE_URL + '/weather/' + locationName, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
            }
        }
    )
}